File: cluster/ClusterClient.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */


import {TypeDBClient} from "../api/TypeDBClient";
import {TypeDBClusterOptions, TypeDBOptions} from "../api/TypeDBOptions";
import {Database} from "../api/database/Database";
import {SessionType} from "../api/TypeDBSession";
import {CoreClient} from "../core/CoreClient";
import {ClusterDatabaseManager} from "./ClusterDatabaseManager";
import {ClusterDatabase} from "./ClusterDatabase";
import {ClusterSession} from "./ClusterSession";
import {FailsafeTask} from "./FailsafeTask";
import {RequestBuilder} from "../common/rpc/RequestBuilder";
import {ErrorMessage} from "../common/errors/ErrorMessage";
import {TypeDBClientError} from "../common/errors/TypeDBClientError";
import {TypeDBClusterClient} from "typedb-protocol/cluster/cluster_service_grpc_pb";
import {ServerManager} from "typedb-protocol/cluster/cluster_server_pb";
import {ChannelCredentials} from "@grpc/grpc-js";
import CLUSTER_UNABLE_TO_CONNECT = ErrorMessage.Client.CLUSTER_UNABLE_TO_CONNECT;

export class ClusterClient implements TypeDBClient.Cluster {

    private _coreClients: { [serverAddress: string]: CoreClient };
    private _typeDBClusterRPCs: { [serverAddress: string]: TypeDBClusterClient };
    private _databaseManagers: ClusterDatabaseManager;
    private _clusterDatabases: { [db: string]: ClusterDatabase };
    private _isOpen: boolean;

    async open(addresses: string[]): Promise<this> {
        const serverAddresses = await this.fetchClusterServers(addresses);
        this._coreClients = {}
        serverAddresses.forEach((addr) => {
            this._coreClients[addr] = new CoreClient(addr);
        });
        this._typeDBClusterRPCs = {};
        serverAddresses.forEach((addr) => {
            this._typeDBClusterRPCs[addr] = new TypeDBClusterClient(addr, ChannelCredentials.createInsecure());
        });

        this._databaseManagers = new ClusterDatabaseManager(this);
        this._clusterDatabases = {};
        this._isOpen = true;
        return this;
    }

    session(database: string, type: SessionType, options: TypeDBClusterOptions = TypeDBOptions.cluster()): Promise<ClusterSession> {
        if (options.readAnyReplica) {
            return this.sessionAnyReplica(database, type, options);
        } else {
            return this.sessionPrimaryReplica(database, type, options);
        }
    }

    private sessionPrimaryReplica(database: string, type: SessionType, options: TypeDBClusterOptions): Promise<ClusterSession> {
        return new OpenSessionFailsafeTask(database, type, options, this).runPrimaryReplica();
    }

    private sessionAnyReplica(database: string, type: SessionType, options: TypeDBClusterOptions): Promise<ClusterSession> {
        return new OpenSessionFailsafeTask(database, type, options, this).runAnyReplica();
    }

    databases(): ClusterDatabaseManager {
        return this._databaseManagers;
    }

    isOpen(): boolean {
        return this._isOpen;
    }

    close(): void {
        if (this._isOpen) {
            this._isOpen = false;
            Object.values(this._coreClients).forEach(client => client.close());
        }
    }

    isCluster(): boolean {
        return true;
    }

    clusterDatabases(): { [db: string]: ClusterDatabase } {
        return this._clusterDatabases;
    }

    clusterMembers(): string[] {
        return Object.keys(this._coreClients);
    }

    coreClient(address: string): TypeDBClient {
        return this._coreClients[address];
    }

    coreClients() {
        return this._coreClients;
    }

    typeDBClusterRPC(address: string): TypeDBClusterClient {
        return this._typeDBClusterRPCs[address];
    }

    private async fetchClusterServers(addresses: string[]): Promise<string[]> {
        for (const address of addresses) {
            const client = new CoreClient(address);
            try {
                console.info(`Fetching list of cluster servers from ${address}...`);
                const grpcClusterClient = new TypeDBClusterClient(address, ChannelCredentials.createInsecure());
                const res = await new Promise<ServerManager.All.Res>((resolve, reject) => {
                    grpcClusterClient.servers_all(RequestBuilder.Cluster.ServerManager.allReq(), (err, res) => {
                        if (err) reject(new TypeDBClientError(err));
                        else resolve(res);
                    });
                });
                const members = res.getServersList().map(x => x.getAddress());
                console.info(`The cluster servers are ${members}`);
                return members;
            } catch (e) {
                console.error(`Fetching cluster servers from ${address} failed.`, e);
            } finally {
                client.close();
            }
        }
        throw new TypeDBClientError(CLUSTER_UNABLE_TO_CONNECT.message(addresses.join(",")));
    }

    asCluster(): TypeDBClient.Cluster {
        return this;
    }
}

class OpenSessionFailsafeTask extends FailsafeTask<ClusterSession> {
    private readonly _type: SessionType;
    private readonly _options: TypeDBClusterOptions;

    constructor(database: string, type: SessionType, options: TypeDBClusterOptions, client: ClusterClient) {
        super(client, database);
        this._type = type;
        this._options = options;
    }

    run(replica: Database.Replica): Promise<ClusterSession> {
        const session = new ClusterSession(this.client, replica.address());
        return session.open(replica.address(), this.database, this._type, this._options);
    }
}

File: cluster/ClusterClient.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.cluster;

import grakn.client.api.GraknClient;
import grakn.client.api.GraknOptions;
import grakn.client.api.GraknSession;
import grakn.client.common.exception.GraknClientException;
import grakn.client.common.rpc.GraknStub;
import grakn.common.collection.Pair;
import grakn.protocol.ClusterServerProto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

import static grakn.client.common.exception.ErrorMessage.Client.CLUSTER_UNABLE_TO_CONNECT;
import static grakn.client.common.exception.ErrorMessage.Client.UNABLE_TO_CONNECT;
import static grakn.client.common.rpc.RequestBuilder.Cluster.ServerManager.allReq;
import static grakn.common.collection.Collections.pair;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toSet;

public class ClusterClient implements GraknClient.Cluster {

    private static final Logger LOG = LoggerFactory.getLogger(ClusterClient.class);

    private final boolean tlsEnabled;
    @Nullable
    private final Path tlsRootCA;
    private final int parallelisation;
    private final Map<String, ClusterNodeClient> clusterNodeClients;
    private final Map<String, GraknStub.Cluster> stubs;
    private final ClusterDatabaseManager databaseMgrs;
    private final ConcurrentMap<String, ClusterDatabase> clusterDatabases;
    private boolean isOpen;

    ClusterClient(Set<String> addresses, boolean tlsEnabled, @Nullable String tlsRootCA) {
        this(addresses, tlsEnabled, tlsRootCA, ClusterNodeClient.calculateParallelisation());
    }

    ClusterClient(Set<String> addresses, boolean tlsEnabled, @Nullable String tlsRootCA, int parallelisation) {
        this.tlsEnabled = tlsEnabled;
        this.tlsRootCA = tlsRootCA != null ? Paths.get(tlsRootCA) : null;
        this.parallelisation = parallelisation;
        clusterNodeClients = fetchServerAddresses(addresses).stream()
                .map(address -> pair(address, ClusterNodeClient.create(address, tlsEnabled, this.tlsRootCA, parallelisation)))
                .collect(toMap(Pair::first, Pair::second));
        stubs = clusterNodeClients.entrySet().stream()
                .map(client -> pair(client.getKey(), GraknStub.cluster(client.getValue().channel())))
                .collect(toMap(Pair::first, Pair::second));
        databaseMgrs = new ClusterDatabaseManager(this);
        clusterDatabases = new ConcurrentHashMap<>();
        isOpen = true;
    }

    public static ClusterClient create(Set<String> addresses, boolean tlsEnabled, @Nullable String tlsRootCA) {
        return new ClusterClient(addresses, tlsEnabled, tlsRootCA);
    }

    public static ClusterClient create(Set<String> addresses, boolean tlsEnabled, @Nullable String tlsRootCA, int parallelisation) {
        return new ClusterClient(addresses, tlsEnabled, tlsRootCA, parallelisation);
    }

    private Set<String> fetchServerAddresses(Set<String> addresses) {
        for (String address : addresses) {
            try (ClusterNodeClient client = ClusterNodeClient.create(address, tlsEnabled, tlsRootCA, parallelisation)) {
                LOG.debug("Fetching list of cluster servers from {}...", address);
                GraknStub.Cluster stub = GraknStub.cluster(client.channel());
                ClusterServerProto.ServerManager.All.Res res = stub.serversAll(allReq());
                Set<String> members = res.getServersList().stream().map(ClusterServerProto.Server::getAddress).collect(toSet());
                LOG.debug("The cluster servers are {}", members);
                return members;
            } catch (GraknClientException e) {
                if (e.getErrorMessage().equals(UNABLE_TO_CONNECT)) {
                    LOG.error("Fetching cluster servers from {} failed.", address);
                } else {
                    throw e;
                }
            }
        }
        throw new GraknClientException(CLUSTER_UNABLE_TO_CONNECT, String.join(",", addresses));
    }

    @Override
    public boolean isOpen() {
        return isOpen;
    }

    @Override
    public ClusterDatabaseManager databases() {
        return databaseMgrs;
    }

    @Override
    public ClusterSession session(String database, GraknSession.Type type) {
        return session(database, type, GraknOptions.cluster());
    }

    @Override
    public ClusterSession session(String database, GraknSession.Type type, GraknOptions options) {
        GraknOptions.Cluster clusterOptions = options.asCluster();
        if (clusterOptions.readAnyReplica().isPresent() && clusterOptions.readAnyReplica().get()) {
            return sessionAnyReplica(database, type, clusterOptions);
        } else {
            return sessionPrimaryReplica(database, type, clusterOptions);
        }
    }

    private ClusterSession sessionPrimaryReplica(String database, GraknSession.Type type, GraknOptions.Cluster options) {
        return openSessionFailsafeTask(database, type, options, this).runPrimaryReplica();
    }

    private ClusterSession sessionAnyReplica(String database, GraknSession.Type type, GraknOptions.Cluster options) {
        return openSessionFailsafeTask(database, type, options, this).runAnyReplica();
    }

    private FailsafeTask<ClusterSession> openSessionFailsafeTask(
            String database, GraknSession.Type type, GraknOptions.Cluster options, ClusterClient client) {
        return new FailsafeTask<ClusterSession>(this, database) {
            @Override
            ClusterSession run(ClusterDatabase.Replica replica) {
                return new ClusterSession(client, replica.address(), database, type, options);
            }
        };
    }

    // TODO: this is not good - we should not pass an internal object to be modified outside of this class
    ConcurrentMap<String, ClusterDatabase> databaseByName() {
        return clusterDatabases;
    }

    Map<String, ClusterNodeClient> clusterNodeClients() {
        return clusterNodeClients;
    }

    Set<String> clusterNodes() {
        return clusterNodeClients.keySet();
    }

    ClusterNodeClient clusterNodeClient(String address) {
        return clusterNodeClients.get(address);
    }

    GraknStub.Cluster stub(String address) {
        return stubs.get(address);
    }

    @Override
    public boolean isCluster() {
        return true;
    }

    @Override
    public Cluster asCluster() {
        return this;
    }

    @Override
    public void close() {
        clusterNodeClients.values().forEach(ClusterNodeClient::close);
        isOpen = false;
    }
}
